be/src/storage/rowset/beta_rowset_writer.cpp

```diff
@@ -425,107 +431,328 @@ Status HorizontalBetaRowsetWriter::_final_merge() {
     MonotonicStopWatch timer;
     timer.start();
 
-    auto schema = vectorized::ChunkHelper::convert_schema_to_format_v2(*_context.tablet_schema);
+    if (CompactionUtils::choose_compaction_algorithm(_context.tablet_schema->num_columns(),
+                                                     config::vertical_compaction_max_columns_per_group,
+                                                     _num_segment) == VERTICAL_COMPACTION) {
+        std::vector<std::vector<uint32_t>> column_groups;
+        CompactionUtils::split_column_into_groups(_context.tablet_schema->num_columns(),
+                                                  _context.tablet_schema->num_key_columns(),
+                                                  config::vertical_compaction_max_columns_per_group, &column_groups);
+
+        auto schema = vectorized::ChunkHelper::convert_schema_to_format_v2(*_context.tablet_schema, column_groups[0]);
+
+        std::vector<vectorized::ChunkIteratorPtr> seg_iterators;
+        seg_iterators.reserve(_num_segment);
+
+        vectorized::SegmentReadOptions seg_options;
+        seg_options.fs = _fs;
+
+        OlapReaderStatistics stats;
+        seg_options.stats = &stats;
+
+        for (int seg_id = 0; seg_id < _num_segment; ++seg_id) {
+            std::string tmp_segment_file =
+                    BetaRowset::segment_temp_file_path(_context.rowset_path_prefix, _context.rowset_id, seg_id);
+            auto segment_ptr = Segment::open(ExecEnv::GetInstance()->tablet_meta_mem_tracker(), _fs, tmp_segment_file,
+                                             seg_id, _context.tablet_schema);
+            if (!segment_ptr.ok()) {
+                LOG(WARNING) << "Fail to open " << tmp_segment_file << ": " << segment_ptr.status();
+                return segment_ptr.status();
+            }
+            if ((*segment_ptr)->num_rows() == 0) {
+                continue;
+            }
+            auto res = (*segment_ptr)->new_iterator(schema, seg_options);
```

Comment:
Why creating an iterator will encounter an end_of_file?

Reply:
changed in https://github.com/StarRocks/starrocks/pull/7333/commits/7c6892471b5664332640c65e28debedf40c09b91